taskchampion/taskchampion/src/server/cloud/gcp.rs

```diff
@@ -0,0 +1,392 @@
+                    let obj = &items[self.next_index];
+                    self.next_index += 1;
+                    let creation = obj.time_created.map(|t| t.unix_timestamp()).unwrap_or(0);
+                    let creation: u64 = creation.try_into().unwrap_or(0);
+                    return Some(Ok(ObjectInfo {
+                        name: obj.name.as_bytes().to_vec(),
+                        creation,
+                    }));
+                } else if result.next_page_token.is_some() {
+                    // Fetch the next page and try again.
+                    if let Err(e) = self.fetch_batch() {
+                        return Some(Err(e));
+                    }
+                    return self.next();
+                }
+            }
+        }
+        None
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use uuid::Uuid;
+
+    /// Make a service if `GCP_TEST_BUCKET` is set, as well as a function to put a unique prefix on
+    /// an object name, so that tests do not interfere with one another.
+    ///
+    /// Set up this bucket with a lifecyle policy to delete objects with age > 1 day. While passing
+    /// tests should correctly clean up after themselves, failing tests may leave objects in the
+    /// bucket.
+    ///
+    /// When the environment variable is not set, this returns false and the test does not run.
+    /// Note that the Rust test runner will still show "ok" for the test, as there is no way to
+    /// indicate anything else.
+    fn make_service() -> Option<(GcpService, impl Fn(&str) -> Vec<u8>)> {
+        let Ok(bucket) = std::env::var("GCP_TEST_BUCKET") else {
+            return None;
+        };
+        let prefix = Uuid::new_v4();
+        Some((GcpService::new(bucket).unwrap(), move |n: &_| {
+            format!("{}-{}", prefix.as_simple(), n).into_bytes()
+        }))
+    }
+
+    #[test]
+    fn put_and_get() {
+        let Some((mut svc, pfx)) = make_service() else {
+            return;
+        };
+        svc.put(&pfx("testy"), b"foo").unwrap();
+        let got = svc.get(&pfx("testy")).unwrap();
+        assert_eq!(got, Some(b"foo".to_vec()));
+
+        // Clean up.
+        svc.del(&pfx("testy")).unwrap();
+    }
+
+    #[test]
+    fn get_missing() {
+        let Some((mut svc, pfx)) = make_service() else {
+            return;
+        };
+        let got = svc.get(&pfx("testy")).unwrap();
+        assert_eq!(got, None);
+    }
+
+    #[test]
+    fn del() {
+        let Some((mut svc, pfx)) = make_service() else {
+            return;
+        };
+        svc.put(&pfx("testy"), b"data").unwrap();
+        svc.del(&pfx("testy")).unwrap();
+        let got = svc.get(&pfx("testy")).unwrap();
+        assert_eq!(got, None);
+    }
+
+    #[test]
+    fn del_missing() {
+        let Some((mut svc, pfx)) = make_service() else {
+            return;
+        };
+
+        // Clean up.
+        svc.del(&pfx("testy")).unwrap();
```

Comment:
My thinking here is that this makes the assertion more explicit by naming its opposite. Disregard if the way you have it is idiomatic rust testing practice.

```suggestion
        svc.del(&pfx("testy")).expect("del returned an Err");
```

Reply:
In a test, the error from "unwrap" is generally sufficient to figure out what went wrong -- it will include a source-code position as well as the display version of the error.

The `.expect` version reads as "this test expects del to return an error" which is in fact the opposite of what's expected. The [docs](https://doc.rust-lang.org/stable/std/result/enum.Result.html#method.expect) suggest using "should", which might be
```suggestion
        svc.del(&pfx("testy")).expect("del should not return an error");
```
which isn't much more informative than `.unwrap()`.

Reply:
I agree that the error when the test fails is perfectly clear, but IMO the purpose of tests is harder to to interpret for the code reader when there is no explicit assertion and, in effect, the assertion is that the test didn't panic. It's not too bad in short tests like this one but in the lengthier ones it seems even more difficult because the code could theoretically panic at any point but the test author only had a limited set of potential panics in mind.

It could also be that this is the idiomatic way to write rust tests and you should leave it as is.

Reply:
Ah, that's true. "Doesn't fail" is always a bit of a hard test to write. Maybe the most idiomatic way to write it is
```rust
assert!(thing.is_ok());
```
I'll make that change.